aten/src/ATen/BatchingRules.cpp

```diff
@@ -0,0 +1,31 @@
+#include <ATen/BatchingRules.h>
+#include <ATen/ATen.h>
+
+namespace at {
+
+std::pair<Tensor,BatchDims> sum_batching_rule(
+    const Tensor& self, BatchDimsRef self_bdims,
+    IntArrayRef dims, bool keepdim, c10::optional<ScalarType> dtype) {
+  // NB: We don't really need to move the batch dims to the front.
+  // One alternative way to do this is to keep them where they are and compute
+  // the required `dims` to reduce over. However, assuming that the batch
+  // dims are at front greatly simplifies the `dims` calculation and moving
+  // them there is relatively cheap.
```

Comment:
Should this "move to the front" simply be backed in as an invariant to the batch dimensions, so you can just assume batch dimensions are in canonical form? What would go wrong otherwise?

Reply:
Ah, I think I remember: the reason you don't do this is because you want to avoid repeatedly transposing the tensor (the way things are setup now, you only need to transpose once, right before use, rather than transposing on very nested vmap). If that's the case, it seems like it another way of explaining `BatchedTensorImpl` is that it records a "delayed" permutation of some sort. In other words, `moveBatchDimsToFront` isn't just a utility function: it is central to the *semantics* of batched tensors.

Although, it is not a complete semantics, since moveBatchDimsToFront really is only going to work for single tensor input operations, right? The multiargument ones are probably getting their own utilities?

Reply:
> Ah, I think I remember: the reason you don't do this is because you want to avoid repeatedly transposing the tensor (the way things are setup now, you only need to transpose once, right before use, rather than transposing on very nested vmap). 

That's part of the reason. Another reason is that there are some ops that perform better when the batch dimension is not at the front (only LSTMs), so the hypothesis is that in the future, we might want some batching rules to not permute dimensions if it impacts performance. 

Alternatively, we could just say this is an invariant for eager-mode vmap and look to the JIT to improve transposing.

> Although, it is not a complete semantics, since moveBatchDimsToFront really is only going to work for single tensor input operations, right? The multiargument ones are probably getting their own utilities?

Yup, the multiargument ones need their own utilities to align dimensions

Reply:
> That's part of the reason. Another reason is that there are some ops that perform better when the batch dimension is not at the front (only LSTMs), so the hypothesis is that in the future, we might want some batching rules to not permute dimensions if it impacts performance.

I guess it doesn't solve the problem if some (most) batching rules move dimensions to the front, but some batching rules don't. If we had a "time, batch, channel" tensor and a vmap over "mul" and then "lstm", "mul" would move the batch dimension to the front first and then the batch dimension would not be in the middle anymore for lstm

Reply:
I mean, it doesn't *really* matter, because the physical layout is the same, and that's what counts, as far as efficiency goes. (Actually, now I wonder how vmap is going to interact with memory layout! If we transpose internal batch dimensions to the left, will they then durably become batch dimensions on the outside?)

Reply:
It seems like it vmap is very related to memory layout, because preserving layout is essential to maintaining semantics for inner batch dimensions.

To be super concrete, let's imagine that we are doing a pointwise operation on a `(1,2,3)` size contiguous tensor, where the 2nd dimension is marked as batch dimension in vmap. At the end of the operation, we should end up with a `(1,2,3)` size contiguous tensor. However, if we transpose, and the pointwise operation doesn't preserve memory layout, you will end up with a physical layout of `(2,1,3)`. In the case of pointwise you can "work around" this by simply not transposing anything, but for example for convolution, you're not going to have a kernel that is going to be able to deal with an internal batch dimension, and so you are going to end up with the "wrong" layout in the end.

cc @VitalyFedyunin, what is the memory layout plan here?

Reply:
> preserving layout is essential to maintaining semantics for inner batch dimensions.

I'm not sure that it is important (or plausible) for vmap to maintain how tensors are strided for inner batch dimensions. The value-statement for vmap is that it makes batching convenient but won't be as fast as manual batching; if users want to fine-tune how the strides of their tensors are for performance reasons, then they're better off doing manual batching.

> but for example for convolution, you're not going to have a kernel that is going to be able to deal with an internal batch dimension, and so you are going to end up with the "wrong" layout in the end.

It looks like in order to support "maintaining strides semantics" for inner batch dimension, then we would need a kernel for each layout here... (or just a generic one that doesn't rely on contiguous or NHWC layout)

Reply:
> I'm not sure that it is important (or plausible) for vmap to maintain how tensors are strided for inner batch dimensions. 

I agree. I think all I am asking for is a semantics. A user asks "what do the strides of vmap look like at the end" and we should be able to answer this question, in much the way we answer the question for most regular operators.